Show left vs. right page layout, even in edit mode BL-4797

hatton opened the pull request:
Switch from relying on css counting for left vs. right styling to  using a class maintained by c#. This allows us to show a page in isolation and still show it correctly in this respect.

TODO: how does this impact epub and bloom reader?
TODO: doesn't yet update display of the page when page is reordered

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/bloombooks/bloomdesktop/1748)
<!-- Reviewable:end -->

andrew-polk commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/BloomExe/Book/Book.cs, line 617 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1748#-KpQTvCmeVWlEdA0-kd1:-KpQTvCmeVWlEdA0-kd2:b-v13mz2) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/6e13a5b2132136aca52ad387ec3114f3490acd34/src/BloomExe/Book/Book.cs#L617)):*
> ```Smalltalk
> 		}
> 
> 		// Reduce repetitive reloading of books when looking up the related "TemplateBook".
> ```

This comment is now misplaced

---

*[src/BloomExe/Book/HtmlDom.cs, line 1445 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1748#-KpQUFTdYHY8ryeTG9Ba:-KpQUFTdYHY8ryeTG9Bb:bpbzq9a) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/6e13a5b2132136aca52ad387ec3114f3490acd34/src/BloomExe/Book/HtmlDom.cs#L1445)):*
> UpdateSideClassOfAllPages

Not a big deal, but if startingIndex is provided, it isn't "all" pages. Could be UpdateSideClassOfPages instead

---

*[src/BloomExe/Book/HtmlDom.cs, line 1457 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1748#-KpQUh7MLkXTPW1WE61O:-KpQUh7MLkXTPW1WE61P:b-xqh710) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/6e13a5b2132136aca52ad387ec3114f3490acd34/src/BloomExe/Book/HtmlDom.cs#L1457)):*
> ```Smalltalk
> 		private static void UpdateSideClass(XmlElement pageDiv, int indexOfPageZeroBased, bool languageIsRightToLeft)
> 		{
> 			RemoveClassesBeginingWith(pageDiv, "side");
> ```

Feels risky. I guess you probably searched less to make sure we aren't using something which starts with "side" already, but we could always introduce one.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1748)*
<!-- Sent from Reviewable.io -->

hatton commented:
Review status: 3 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/BloomExe/Book/Book.cs, line 617 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1748#-KpQTvCmeVWlEdA0-kd1:-KpQatTEGag68MhHR24w:b28nszh) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/6e13a5b2132136aca52ad387ec3114f3490acd34/src/BloomExe/Book/Book.cs#L617)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

This comment is now misplaced
</blockquote></details>

Whole method is now gone, as the result of some new work tracking down the page-display update bug this morning.

---

*[src/BloomExe/Book/HtmlDom.cs, line 1445 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1748#-KpQUFTdYHY8ryeTG9Ba:-KpQb4wmsoexSEa0DugB:b-u7h2tl) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/6e13a5b2132136aca52ad387ec3114f3490acd34/src/BloomExe/Book/HtmlDom.cs#L1445)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> UpdateSideClassOfAllPages

Not a big deal, but if startingIndex is provided, it isn't "all" pages. Could be UpdateSideClassOfPages instead
</blockquote></details>

It is still all the pages in the dom that is supplied. No pages are ever skipped. The startingIndex, rather, is the location of this page in the book, which you have to tell us if that dom doesn't represent the whole book.

Added more comments to clarify this.

---

*[src/BloomExe/Book/HtmlDom.cs, line 1457 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1748#-KpQUh7MLkXTPW1WE61O:-KpQbzjBCi83XnSpEpZd:b-pqcv78) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/6e13a5b2132136aca52ad387ec3114f3490acd34/src/BloomExe/Book/HtmlDom.cs#L1457)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Feels risky. I guess you probably searched less to make sure we aren't using something which starts with "side" already, but we could always introduce one.
</blockquote></details>

changed to "side-". 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1748)*
<!-- Sent from Reviewable.io -->

andrew-polk commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/BloomExe/Book/Book.cs, line 617 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1748#-KpQTvCmeVWlEdA0-kd1:-KpQxsMpQfAn7s9U78bR:b-nengub) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/6e13a5b2132136aca52ad387ec3114f3490acd34/src/BloomExe/Book/Book.cs#L617)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

Whole method is now gone, as the result of some new work tracking down the page-display update bug this morning.
</blockquote></details>

I miscommunicated. This comment was originally for the code below it. So the net result now is a comment has been removed.

With the removal of this call, I don't think there are any calls to UpdateSideClassOfAllPages which use the startingIndex parameter.

---

*[src/BloomExe/Book/HtmlDom.cs, line 1447 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1748#-KpQz9yvLZtR3bluQhLV:-KpQz9yvLZtR3bluQhLW:b46ltjj) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/e107f23187575290b1c219f6407822216442940b/src/BloomExe/Book/HtmlDom.cs#L1447)):*
> ```Smalltalk
> 		/// <param name="startingIndex">If this dom is only a subset of the book, then you need to provide this
> 		/// to tell us what the first page index is, of this subset of the full dom.</param>
> 		public void UpdateSideClassOfAllPages(bool languageIsRightToLeft, int startingIndex = 0)
> ```

Worth adding any tests for this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1748)*
<!-- Sent from Reviewable.io -->

hatton commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/BloomExe/Book/HtmlDom.cs, line 1447 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1748#-KpQz9yvLZtR3bluQhLV:-KpRChb7ZFE2RuXm77b_:br8xbje) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/e107f23187575290b1c219f6407822216442940b/src/BloomExe/Book/HtmlDom.cs#L1447)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Worth adding any tests for this?
</blockquote></details>

not yet, too trivial. But maybe after this gets merged and I integrate the page numbering, which is more complicated.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1748)*
<!-- Sent from Reviewable.io -->

hatton commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/BloomExe/Book/Book.cs, line 617 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1748#-KpQTvCmeVWlEdA0-kd1:-KpRCtubG_keJXYQjRZ2:b-vk8iow) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/6e13a5b2132136aca52ad387ec3114f3490acd34/src/BloomExe/Book/Book.cs#L617)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

I miscommunicated. This comment was originally for the code below it. So the net result now is a comment has been removed.

With the removal of this call, I don't think there are any calls to UpdateSideClassOfAllPages which use the startingIndex parameter.
</blockquote></details>

yah, it's actually gone in the next batch of work with page numbers.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1748)*
<!-- Sent from Reviewable.io -->
